TASK-87: Messages Page UX/Styling Cleanup

cristian-aldea opened the pull request:
closes #315

To test this PR, make sure the following points are addressed:
- 87.1 Remove chatbox autofill
    - I didn't get to test this one, but looking online I just needed to set autoComplete to off on the form
- 87.2 Don’t allow empty spaces without photos to be sent 
- 87.3 Add username to chat bubbles
- 87.4 Text input area should grow to a certain size if the message is really long (keeping in mind the backend varchar length)
    - I've gone through several attempts to do this, I thought there would be a native solution to this, but I guess not ¯\\\_(ツ)\_/¯. The "textarea" is actually a div, so unfortunately you can't treat it as a normal form element :/. This isn't a big issue, just something to keep in mind
- 87.5 Text message limit should be shown (or figure out a way to send long messages)
    - You'll see this as a `xx/512` once you approach the limit. you'll see a negative number once you go over the limit.
- 87.6 Profile pics are clickable and should redirect to their respective profile page
- 87.7 Remove bottom nav when keyboard is showing (mobile)
    - My implementation doesn't take into account users who resize their window to try and break this, so don't @ me about that.

cristian-aldea committed:
fix(chat input): set autocomplete to "off"

ref #315

cristian-aldea committed:
fix(chat input): Add robust check for whitespace

Also, the following misc problems I saw were fixed:
- moved .chat-bubble-image to correct file
- added border-radius to .chat-bubble-image
- Fixed chat input image to have object-fit and some margin on the right
- fixed missing check for errors in message in ChatInput
- Added button class to icon you press to upload photo for feedback
- add toast error message when message fails to be sent
- add trim to chat message. This should be added everywhere honestly

ref #315

cristian-aldea committed:
fix(chat bubble): Add user to chat bubble

Also, added button class to CustomFileUpload to get hover/click feedback

ref #315

cristian-aldea committed:
fix(chat input): grow chat input as user types

Also, reworked a bunch of stuff:
- added length to post message, both in db and form validation
- reworked a lot of styling
- created new MullTextArea component
-fixed eventcard bug where event.id was used to fetch image instead of event.image.id
- set white-space of chat bubble to pre-wrap so that new lines written in chatinput are preserved
- fixed svgIcon placement in custom text input to be perfectly vertically centered
- removed instances of textarea in create event and replaced with MullTextArea

ref #315

cristian-aldea committed:
fix(chat input): Fix issues with growing chat input

- Submit form on enter, but not shift enter
- scroll page as input height increases

ref #315

cristian-aldea committed:
fix(chat input): add indicator for post length limit

ref #315

cristian-aldea committed:
fix(chat input): Hide bot nav on focus

ref #315

cristian-aldea committed:
fix(chat input): Fix issue where you would need to press the submit button twice

ref #315

cristian-aldea committed:
fix: Adjust error messages to be accurate

ref #315

cristian-aldea committed:
fix: Address lint/test/e2e issues

ref #315

osahtout commented on apps/mull-ui/src/app/pages/messages/event-chat/event-chat.tsx:
```diff
@@ -99,7 +100,11 @@ export const EventChat = ({ history, channelName, restrictChatInput }: EventChat
+        ? Yup.string().optional()
+        : Yup.string()
+            .required()
+            .test('no-whitespace', "Message can't be empty", (value) => !/^\s*$/.test(value)),
```
Message isn't really empty since it does have a space character. Not having any characters will also show the user "message is a required field" maybe just not have any warnings for the user

cristian-aldea replied:
> maybe just not have any warnings for the user

well no they can't send a message that is empty or just has whitespace. I'll adjust the error message

osahtout replied:
Well because now there's two different error message that trying to say the same thing.

When you type a space, the message isn't empty.

I don't think showing an error message for sending an empty message or a space.

cristian-aldea replied:
Sure, we could have it just not do anything. This actually also solves another issue where the error message blocks off messages. will do

osahtout commented on apps/mull-ui/src/app/components/chat-bubble/chat-bubble.tsx:
```diff
@@ -1,26 +1,31 @@
+    <div className="chat-bubble-container">
+      <div className="chat-bubble-header">
+        <p>{chatDate}</p>
+        <p className="chat-bubble-header-username">{user.name}</p>
```
This is cause some styling issues and doesn't look quite right
![image](https://user-images.githubusercontent.com/44760251/113918503-8406ab80-97b0-11eb-86e1-c918d6167b95.png)
![image](https://user-images.githubusercontent.com/44760251/113918700-bf08df00-97b0-11eb-8224-346b83c56745.png)


I think it should just be the first name and should be at the start of the bubble

also if your the sender, I don't think it should your name

cristian-aldea replied:
> I think it should just be the first name

well we don't have a first name field so I can't know what their first name is for sure.

> also if your the sender, I don't think it should your name

Not sure what you mean

cristian-aldea replied:
I'll try and see if I can address the overlap thou

osahtout replied:
At least until the first white space or trim down the name. There should be a limit

just like when you're the sender the bubble is green, it should not show your name as you know who you are

cristian-aldea replied:
> At least until the first white space or trim down the name

I don't think its safe to make assumptions like that. The username might be `The recycler` and splitting on whitespace would give "The", which looks weird.

I'll make the font smaller, and remove the user text for the current user

cristian-aldea replied:
Nevermind, came up with some clever styling :)
https://user-images.githubusercontent.com/23544999/113974814-8eef2980-980c-11eb-8420-61a281f6e715.mp4

osahtout commented on apps/mull-ui/src/app/components/chat-input/chat-input.tsx:
```diff
@@ -89,6 +88,11 @@ export const ChatInput = ({
         />
+        {LIMITS.POST_MESSAGE - formik.values.message.length < 100 ? (
+          <div className="chat-input-limit-message">
+            {LIMITS.POST_MESSAGE - formik.values.message.length}/{LIMITS.POST_MESSAGE}
```
this will show a negative number if it's over the limit (i.e -1/512 instead of 513/512)

RGPosadas replied:
To add on to this, the text limit stays until the user starts typing in the chat box again
![image](https://user-images.githubusercontent.com/33991121/113925749-61c55b80-97b9-11eb-9590-77c577b9dfb7.png)

cristian-aldea replied:
The negative number is intentional, a few apps do that

> the text limit stays until the user starts typing in the chat box again

Will try and fix this

osahtout replied:
Can you show which apps do that, i'm very curious

cristian-aldea replied:
twitter for one
![image](https://user-images.githubusercontent.com/23544999/113975127-0cb33500-980d-11eb-8c8e-de3b9af9341d.png)

cristian-aldea replied:
Fixed issue that ritchelle mentionned. Was due to some desync between form values and textcontent of div.

osahtout reviewed (commented):
https://user-images.githubusercontent.com/44760251/113920603-fb3d3f00-97b2-11eb-952c-981ad5401a01.MP4

there are some blocking issues in this PR. 

I’m also not sure if I welcome the changes that’s isn’t part of this ticket. It makes it difficult to review it.

RGPosadas commented on apps/mull-ui/src/app/components/chat-bubble/chat-bubble.tsx:
```diff
@@ -1,26 +1,31 @@
+    <div className="chat-bubble-container">
+      <div className="chat-bubble-header">
+        <p>{chatDate}</p>
+        <p className="chat-bubble-header-username">{user.name}</p>
```
When looking at posts from a different user, the name should be on the left
![image](https://user-images.githubusercontent.com/33991121/113927017-ef557b00-97ba-11eb-93e8-db6b0a305c6f.png)

cristian-aldea replied:
Good catch, forgot about that

RGPosadas reviewed (commented):
Omar spotted what I had seen as well, so I just 👍 'd and added to them. All of the subtasks have indeed been met, however they do need some tweaking.  Tests + e2e + linting all pass.

**Subtasks**
- [x] 87.1 Remove chatbox autofill
- [x] 87.2 Don’t allow empty spaces without photos to be sent 
- [x] 87.3 Add username to chat bubbles
- [x] 87.4 Text input area should grow to a certain size if the message is really long (keeping in mind the backend varchar length)
- [x] 87.5 Text message limit should be shown (or figure out a way to send long messages)
- [x] 87.6 Profile pics are clickable and should redirect to their respective profile page
- [x] 87.7 Remove bottom nav when keyboard is showing (mobile)

**Acceptance Criteria**
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every new component and function
- [x] The application should still work with seeded data

cristian-aldea commented:
> there are some blocking issues in this PR.

Sorry about that its most likely a quick adjustment that needs to be done to some calculations I do to figure out how much bottom margin to add to the chat message list.

> I’m also not sure if I welcome the changes that’s isn’t part of this ticket. It makes it difficult to review it.

I mostly just refined the styling in some parts. Just gotta check if everything works as before. Check commit messages for details

osahtout commented:
> I mostly just refined the styling in some parts. Just gotta check if everything works as before. Check commit messages for details

Sure, but we were spliting the user stories and task into more granular task to make easier for us to implement and review

cristian-aldea committed:
fix(chat input): Remove error message from chat input

- The chat input will now simply clear the input text and fail the form submission silently. Making it more similar to other apps like messenger
- Fixed annoying behavior where chat would flicker in size after pressing enter to submit the form. This was caused by the form submit check being in onkeydown and not in onInput
- Fixed TS compiler warning

ref #315

cristian-aldea committed:
fix(chat input): Employ better styling for chat bubble username

ref #315

cristian-aldea committed:
fix(chat input): Improve styling of chat max char error

ref #315

cristian-aldea committed:
fix(chat input): maybe fixed ios margin issue

ref #315

cristian-aldea commented:
Addressed all issues. Please have another look when you can @osahtout @RGPosadas

cristian-aldea committed:
fix(chat input): fix scroll sometimes not working

ref #315

cristian-aldea commented:
>  IMG_4208.MP4 

Let me know if I fixed the ios issue you had, should be good now.

RGPosadas commented on apps/mull-ui/src/app/pages/messages/event-chat/event-chat.tsx:
```diff
@@ -130,20 +151,75 @@ export const EventChat = ({ history, channelName, restrictChatInput }: EventChat
+      updateContainerStyle();
+    }
+  };
+  const inputOnBlur = () => {
```
Had some issue with the bottom nav bar not coming back if I don't click out of the event chat input

https://user-images.githubusercontent.com/33991121/114085066-dca47a00-987e-11eb-93a0-ce150b630bae.mov

cristian-aldea replied:
Should be fixable by force blurring the chat input once you click enter. I'll see what I can do

RGPosadas reviewed (approved):
Great work, Cristian! Just had one small comment

**Subtasks**
- [x] 87.1 Remove chatbox autofill
- [x] 87.2 Don’t allow empty spaces without photos to be sent 
- [x] 87.3 Add username to chat bubbles
- [x] 87.4 Text input area should grow to a certain size if the message is really long (keeping in mind the backend varchar length)
- [x] 87.5 Text message limit should be shown (or figure out a way to send long messages)
- [x] 87.6 Profile pics are clickable and should redirect to their respective profile page
- [x] 87.7 Remove bottom nav when keyboard is showing (mobile)

**Acceptance Criteria**
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every new component and function
- [x] The application should still work with seeded data

cristian-aldea committed:
fix: Integrate changes to direct messages page

ref #315

cindyslittleplanet reviewed (approved):
LGTM! 🚀 

**Subtasks**
- [x] 87.1 Remove chatbox autofill
- [x] 87.2 Don’t allow empty spaces without photos to be sent 
- [x] 87.3 Add username to chat bubbles
- [x] 87.4 Text input area should grow to a certain size if the message is really long (keeping in mind the backend varchar length)
- [x] 87.5 Text message limit should be shown (or figure out a way to send long messages)
- [x] 87.6 Profile pics are clickable and should redirect to their respective profile page
- [x] 87.7 Remove bottom nav when keyboard is showing (mobile)

**Acceptance Criteria**
- [x] Documentation present where necessary
- [x] Unit and UI tests are implemented for every new component and function
- [x] The application should still work with seeded data
